subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/MavenComponentParser.java

```diff
@@ -248,42 +242,56 @@ private static Stream<? extends SoftwareComponentVariant> createSortedVariantsSt
     private static class MavenDependencyFactory {
 
         private final VariantWarningCollector warnings;
-        private final VariantDependencyResolver dependencyResolver;
+        private final VariantDependencyResolver variantDependencyResolver;
+        private final ComponentDependencyResolver componentDependencyResolver;
+        private final VersionRangeMapper versionRangeMapper;
 
         private final String scope;
         private final boolean optional;
         private final Set<ExcludeRule> globalExcludes;
 
         public MavenDependencyFactory(
             VariantWarningCollector warnings,
-            VariantDependencyResolver dependencyResolver,
+            VariantDependencyResolver variantDependencyResolver,
+            ComponentDependencyResolver componentDependencyResolver,
+            VersionRangeMapper versionRangeMapper,
             String scope,
             boolean optional,
             Set<ExcludeRule> globalExcludes
         ) {
             this.warnings = warnings;
-            this.dependencyResolver = dependencyResolver;
+            this.variantDependencyResolver = variantDependencyResolver;
+            this.componentDependencyResolver = componentDependencyResolver;
+            this.versionRangeMapper = versionRangeMapper;
             this.scope = scope;
             this.optional = optional;
             this.globalExcludes = globalExcludes;
         }
 
         private void convertDependency(ModuleDependency dependency, Consumer<MavenDependency> collector) {
+
+            if (!dependency.getAttributes().isEmpty()) {
+                warnings.addUnsupported(String.format("%s:%s:%s declared with Gradle attributes", dependency.getGroup(), dependency.getName(), dependency.getVersion()));
+            }
+            if (!dependency.getRequestedCapabilities().isEmpty()) {
+                warnings.addUnsupported(String.format("%s:%s:%s declared with Gradle capabilities", dependency.getGroup(), dependency.getName(), dependency.getVersion()));
+            }
```

Comment:
❓ Will these restrictions eventually go away when we're able to map these to the correct variant as long as that variant is mapped to a different set of coordinates?

Reply:
We're likely going to do some follow-up work to rethink how we want to emit these warnings. As it stands, these warnings are not great, since even variants that _dont_ have attributes could still be mapped to different coordinates. 

In reality, looking at these values to determine if we should warn is likely not what we want. TBH we can't even tell if a variant _would_ be mapped to different coordinates without doing a resolution. 

Really what this means is that we should deprecate publishing without doing a resolution, which is a far way away. 

Right now, this PR has the same warning behavior as before we made any changes so I think we can keep this for now, but we will need to revisit soon